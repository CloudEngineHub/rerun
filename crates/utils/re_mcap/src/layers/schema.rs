use std::sync::Arc;

use arrow::{
    array::{MapBuilder, StringBuilder},
    error::ArrowError,
};
use re_chunk::{Chunk, RowId, TimePoint};
use re_types::{ArchetypeBuilder, ComponentBatch as _, SerializedComponentBatch, components};
use re_types_core::AsComponents as _;

use crate::Error;

use super::{Layer, LayerIdentifier};

/// Extracts a static summary of channel and schema information.
///
/// Can be used to get an overview over the contents of an MCAP file.
#[derive(Debug, Default)]
pub struct McapSchemaLayer;

impl Layer for McapSchemaLayer {
    fn identifier() -> LayerIdentifier {
        "schema".into()
    }

    fn process(
        &mut self,
        _mcap_bytes: &[u8],
        summary: &mcap::Summary,
        emit: &mut dyn FnMut(Chunk),
    ) -> Result<(), Error> {
        for channel in summary.channels.values() {
            let mut components = from_channel(channel)?;
            if let Some(schema) = channel.schema.as_ref() {
                components.extend(from_schema(schema)?);
            }

            let chunk = Chunk::builder(channel.topic.as_str())
                .with_archetype(RowId::new(), TimePoint::STATIC, &components)
                .build()?;
            emit(chunk);
        }

        Ok(())
    }
}

fn from_channel(
    channel: &Arc<::mcap::Channel<'_>>,
) -> Result<Vec<SerializedComponentBatch>, ArrowError> {
    use arrow::array::{StringArray, UInt16Array};

    let ::mcap::Channel {
        id,
        topic,
        schema: _, // handled by `fn from_schema` instead
        message_encoding,
        metadata,
    } = channel.as_ref();

    let key_builder = StringBuilder::new();
    let val_builder = StringBuilder::new();

    let mut builder = MapBuilder::new(None, key_builder, val_builder);

    for (key, val) in metadata {
        builder.keys().append_value(key);
        builder.values().append_value(val);
        builder.append(true)?;
    }

    let metadata = builder.finish();
    let archetype = "rerun.mcap.Channel";

    let channel = ArchetypeBuilder::new(archetype)
        .with_field("id", Arc::new(UInt16Array::from(vec![*id])))
        .with_field("topic", Arc::new(StringArray::from(vec![topic.clone()])))
        .with_field("metadata", Arc::new(metadata))
        .with_field(
            "message_encoding",
            Arc::new(StringArray::from(vec![message_encoding.clone()])),
        );

    // TODO(nick): Now that we have the nicer archetypes builder just pass that through instead of batches
    Ok(channel.as_serialized_batches())
}

fn from_schema(
    schema: &Arc<::mcap::Schema<'_>>,
) -> Result<Vec<SerializedComponentBatch>, re_types::SerializationError> {
    use arrow::array::{StringArray, UInt16Array};

    let ::mcap::Schema {
        id,
        name,
        encoding,
        data,
    } = schema.as_ref();

    let blob = components::Blob(data.clone().into_owned().into());

    // Adds a field of arbitrary data to this archetype.
    let archetype = "rerun.mcap.Schema";

    let schema = ArchetypeBuilder::new(archetype)
        .with_field("id", Arc::new(UInt16Array::from(vec![*id])))
        .with_field("name", Arc::new(StringArray::from(vec![name.clone()])))
        .with_field("data", blob.to_arrow()?)
        .with_field(
            "encoding",
            Arc::new(StringArray::from(vec![encoding.clone()])),
        );
    Ok(schema.as_serialized_batches())
}
