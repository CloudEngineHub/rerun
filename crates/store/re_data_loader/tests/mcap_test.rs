#[cfg(test)]
mod tests {
    use std::fs;

    use prost::Message;
    use prost_reflect::{
        DescriptorPool, DynamicMessage, FieldDescriptor, Kind, MessageDescriptor, Value,
    };
    use prost_types::FileDescriptorSet;
    use re_data_loader::{DataLoaderSettings, loader_mcap::load_mcap};
    use re_mcap::layers::SelectedLayers;

    #[allow(clippy::derive_partial_eq_without_eq)]
    #[derive(Clone, PartialEq, ::prost::Message)]
    pub struct TestProtoSchema {
        #[prost(double, tag = "1")]
        pub x: f64,
        #[prost(double, tag = "2")]
        pub y: f64,
        #[prost(double, tag = "3")]
        pub z: f64,
    }

    fn generate_proto_mcap() -> Vec<u8> {
        let mut buffer = Vec::new();
        {
            let mut writer = mcap::Writer::new(std::io::Cursor::new(&mut buffer))
                .expect("Failed to create MCAP writer");

            // Register a schema
            let schema_name = "TestProtoSchema";
            let schema_encoding = "protobuf";
            // let schema_data = br#"
            // syntax = "proto3";
            // message TestProtoSchema {
            //     double x = 1;
            //     double y = 2;
            //     double z = 3;
            // }"#;

            let descriptor_data =
                fs::read("tests/assets/file_descriptor_set").expect("Failed to read descriptors");
            let file_descriptor_set = FileDescriptorSet::decode(&descriptor_data[..])
                .expect("Failed to decode descriptors");
            let schema_data = file_descriptor_set.encode_to_vec();

            let schema_id = writer
                .add_schema(schema_name, schema_encoding, &schema_data)
                .expect("Failed to add schema");

            // let mut schema_raw = (schema_data.len() as u32).to_le_bytes().to_vec();
            // // schema_raw.extend_from_slice(schema_data);

            // let schema_id = writer
            //     .add_schema(schema_name, schema_encoding, schema_data)
            // .expect("Failed to add schema");

            // Register a channel
            let channel_topic = "/test/points";
            let message_encoding = "protobuf";
            let channel_id = writer
                .add_channel(
                    schema_id,
                    channel_topic,
                    message_encoding,
                    &std::collections::BTreeMap::new(),
                )
                .expect("Failed to add channel");

            // Write some test messages
            for i in 0..5 {
                let timestamp = 1000000000 + i * 100000000; // nanoseconds

                let test_proto_schema = TestProtoSchema {
                    x: 1.0,
                    y: 2.0,
                    z: 3.0,
                };
                let mut message_data = Vec::new();
                test_proto_schema
                    .encode(&mut message_data)
                    .expect("Failed to encode message");

                // // Simple CDR-encoded point data (x=i, y=i*2, z=i*3)
                // let mut message_data = Vec::new();
                // message_data.extend_from_slice(&(i as f64).to_le_bytes()); // x
                // message_data.extend_from_slice(&((i * 2) as f64).to_le_bytes()); // y
                // message_data.extend_from_slice(&((i * 3) as f64).to_le_bytes()); // z

                let message_header = mcap::records::MessageHeader {
                    channel_id,
                    sequence: i as u32,
                    log_time: timestamp,
                    publish_time: timestamp,
                };
                writer
                    .write_to_known_channel(&message_header, &message_data)
                    .expect("Failed to write message");
            }

            writer.finish().expect("Failed to finish writer");
        }
        buffer
    }

    #[test]
    fn test_load_generated_mcap_file() {
        // let mcap_data = generate_ros2_mcap();
        let mcap_data = generate_proto_mcap();
        let (tx, rx) = std::sync::mpsc::channel();
        let settings = DataLoaderSettings::recommended("test");
        load_mcap(&mcap_data, &settings, &tx, SelectedLayers::All).expect("Failed to load MCAP");
        drop(tx);

        while let Ok(res) = rx.recv() {
            println!("Generated MCAP result: {res:?}");
        }
    }
}
