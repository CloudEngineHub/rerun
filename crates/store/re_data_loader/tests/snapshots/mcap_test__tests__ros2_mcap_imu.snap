---
source: crates/store/re_data_loader/tests/mcap_test.rs
expression: chunks
---
[
    Chunk {
        id: ChunkId(
            0000000000661EFDf2e3b19f7c045f15,
        ),
        entity_path: /test/imu,
        heap_size_bytes: 1616,
        is_sorted: true,
        row_ids: FixedSizeBinaryArray<16>
        [
          [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
        ],
        ],
        timelines: {
            "log_time": TimeColumn {
                timeline: Timeline {
                    name: "log_time",
                    typ: TimestampNs,
                },
                times: ScalarBuffer(
                    [
                        1000000000,
                    ],
                ),
                is_sorted: true,
                time_range: AbsoluteTimeRange {
                    min: TimeInt(1 000 000 000),
                    max: TimeInt(1 000 000 000),
                },
            },
            "publish_time": TimeColumn {
                timeline: Timeline {
                    name: "publish_time",
                    typ: TimestampNs,
                },
                times: ScalarBuffer(
                    [
                        1000000000,
                    ],
                ),
                is_sorted: true,
                time_range: AbsoluteTimeRange {
                    min: TimeInt(1 000 000 000),
                    max: TimeInt(1 000 000 000),
                },
            },
        },
        components: ChunkComponents(
            {
                ComponentDescriptor {
                    archetype: Some(
                        "rerun.archetypes.McapMessage",
                    ),
                    component: "McapMessage:data",
                    component_type: Some(
                        "rerun.components.Blob",
                    ),
                }: ListArray
                [
                  ListArray
                [
                  PrimitiveArray<UInt8>
                [
                  0,
                  1,
                  0,
                  0,
                  123,
                  0,
                  0,
                  0,
                  45,
                  0,
                  ...296 elements...,
                  0,
                  64,
                  0,
                  0,
                  0,
                  0,
                  0,
                  0,
                  0,
                  64,
                ],
                ],
                ],
            },
        ),
    },
    Chunk {
        id: ChunkId(
            0000000000661EFDf2e3b19f7c045f16,
        ),
        entity_path: /__properties,
        heap_size_bytes: 384,
        is_sorted: true,
        row_ids: FixedSizeBinaryArray<16>
        [
          [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
        ],
        ],
        timelines: {},
        components: ChunkComponents(
            {
                ComponentDescriptor {
                    archetype: Some(
                        "rerun.archetypes.RecordingInfo",
                    ),
                    component: "RecordingInfo:start_time",
                    component_type: Some(
                        "rerun.components.Timestamp",
                    ),
                }: ListArray
                [
                  PrimitiveArray<Int64>
                [
                  1000000000,
                ],
                ],
            },
        ),
    },
    Chunk {
        id: ChunkId(
            0000000000661EFDf2e3b19f7c045f17,
        ),
        entity_path: /test/imu,
        heap_size_bytes: 42392,
        is_sorted: true,
        row_ids: FixedSizeBinaryArray<16>
        [
          [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
        ],
        ],
        timelines: {
            "log_time": TimeColumn {
                timeline: Timeline {
                    name: "log_time",
                    typ: TimestampNs,
                },
                times: ScalarBuffer(
                    [
                        1000000000,
                    ],
                ),
                is_sorted: true,
                time_range: AbsoluteTimeRange {
                    min: TimeInt(1 000 000 000),
                    max: TimeInt(1 000 000 000),
                },
            },
            "publish_time": TimeColumn {
                timeline: Timeline {
                    name: "publish_time",
                    typ: TimestampNs,
                },
                times: ScalarBuffer(
                    [
                        1000000000,
                    ],
                ),
                is_sorted: true,
                time_range: AbsoluteTimeRange {
                    min: TimeInt(1 000 000 000),
                    max: TimeInt(1 000 000 000),
                },
            },
            "timestamp": TimeColumn {
                timeline: Timeline {
                    name: "timestamp",
                    typ: TimestampNs,
                },
                times: ScalarBuffer(
                    [
                        123000000045,
                    ],
                ),
                is_sorted: true,
                time_range: AbsoluteTimeRange {
                    min: TimeInt(123 000 000 045),
                    max: TimeInt(123 000 000 045),
                },
            },
        },
        components: ChunkComponents(
            {
                ComponentDescriptor {
                    archetype: Some(
                        "rerun.archetypes.Scalars",
                    ),
                    component: "Scalars:scalars",
                    component_type: Some(
                        "rerun.components.Scalar",
                    ),
                }: ListArray
                [
                  PrimitiveArray<Float64>
                [
                  11.0,
                  12.0,
                  13.0,
                  21.0,
                  22.0,
                  23.0,
                ],
                ],
                ComponentDescriptor {
                    archetype: Some(
                        "sensor_msgs.msg.Imu",
                    ),
                    component: "sensor_msgs.msg.Imu:linear_acceleration_covariance",
                    component_type: None,
                }: ListArray
                [
                  PrimitiveArray<Float64>
                [
                  2.0,
                  2.0,
                  2.0,
                  2.0,
                  2.0,
                  2.0,
                  2.0,
                  2.0,
                  2.0,
                ],
                ],
                ComponentDescriptor {
                    archetype: Some(
                        "sensor_msgs.msg.Imu",
                    ),
                    component: "sensor_msgs.msg.Imu:orientation",
                    component_type: None,
                }: ListArray
                [
                  PrimitiveArray<Float64>
                [
                  0.5,
                  0.1,
                  0.7,
                  1.125,
                ],
                ],
                ComponentDescriptor {
                    archetype: Some(
                        "sensor_msgs.msg.Imu",
                    ),
                    component: "sensor_msgs.msg.Imu:orientation_covariance",
                    component_type: None,
                }: ListArray
                [
                  PrimitiveArray<Float64>
                [
                  1.0,
                  2.0,
                  3.0,
                  4.0,
                  5.0,
                  6.0,
                  7.0,
                  8.0,
                  9.0,
                ],
                ],
                ComponentDescriptor {
                    archetype: Some(
                        "sensor_msgs.msg.Imu",
                    ),
                    component: "sensor_msgs.msg.Imu:angular_velocity_covariance",
                    component_type: None,
                }: ListArray
                [
                  PrimitiveArray<Float64>
                [
                  1.0,
                  1.0,
                  1.0,
                  1.0,
                  1.0,
                  1.0,
                  1.0,
                  1.0,
                  1.0,
                ],
                ],
            },
        ),
    },
    Chunk {
        id: ChunkId(
            0000000000661EFDf2e3b19f7c045f18,
        ),
        entity_path: /test/imu,
        heap_size_bytes: 476,
        is_sorted: true,
        row_ids: FixedSizeBinaryArray<16>
        [
          [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
        ],
        ],
        timelines: {},
        components: ChunkComponents(
            {
                ComponentDescriptor {
                    archetype: Some(
                        "rerun.archetypes.SeriesLines",
                    ),
                    component: "SeriesLines:names",
                    component_type: Some(
                        "rerun.components.Name",
                    ),
                }: ListArray
                [
                  StringArray
                [
                  "gyroscope/x",
                  "gyroscope/y",
                  "gyroscope/z",
                  "accelerometer/x",
                  "accelerometer/y",
                  "accelerometer/z",
                ],
                ],
            },
        ),
    },
    Chunk {
        id: ChunkId(
            0000000000661EFDf2e3b19f7c045f19,
        ),
        entity_path: /test/imu,
        heap_size_bytes: 2482,
        is_sorted: true,
        row_ids: FixedSizeBinaryArray<16>
        [
          [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
        ],
        ],
        timelines: {},
        components: ChunkComponents(
            {
                ComponentDescriptor {
                    archetype: Some(
                        "rerun.archetypes.McapSchema",
                    ),
                    component: "McapSchema:encoding",
                    component_type: Some(
                        "rerun.components.Text",
                    ),
                }: ListArray
                [
                  StringArray
                [
                  "ros2msg",
                ],
                ],
                ComponentDescriptor {
                    archetype: Some(
                        "rerun.archetypes.McapChannel",
                    ),
                    component: "McapChannel:metadata",
                    component_type: Some(
                        "rerun.components.KeyValuePairs",
                    ),
                }: ListArray
                [
                  ListArray
                [
                  StructArray
                -- validity:
                [
                ]
                [
                -- child 0: "first" (Utf8)
                StringArray
                [
                ]
                -- child 1: "second" (Utf8)
                StringArray
                [
                ]
                ],
                ],
                ],
                ComponentDescriptor {
                    archetype: Some(
                        "rerun.archetypes.McapSchema",
                    ),
                    component: "McapSchema:name",
                    component_type: Some(
                        "rerun.components.Text",
                    ),
                }: ListArray
                [
                  StringArray
                [
                  "sensor_msgs/msg/Imu",
                ],
                ],
                ComponentDescriptor {
                    archetype: Some(
                        "rerun.archetypes.McapChannel",
                    ),
                    component: "McapChannel:message_encoding",
                    component_type: Some(
                        "rerun.components.Text",
                    ),
                }: ListArray
                [
                  StringArray
                [
                  "cdr",
                ],
                ],
                ComponentDescriptor {
                    archetype: Some(
                        "rerun.archetypes.McapSchema",
                    ),
                    component: "McapSchema:data",
                    component_type: Some(
                        "rerun.components.Blob",
                    ),
                }: ListArray
                [
                  ListArray
                [
                  PrimitiveArray<UInt8>
                [
                  105,
                  103,
                  110,
                  111,
                  114,
                  101,
                  100,
                  95,
                  115,
                  99,
                  ...6 elements...,
                  101,
                  115,
                  99,
                  114,
                  105,
                  112,
                  116,
                  105,
                  111,
                  110,
                ],
                ],
                ],
                ComponentDescriptor {
                    archetype: Some(
                        "rerun.archetypes.McapChannel",
                    ),
                    component: "McapChannel:topic",
                    component_type: Some(
                        "rerun.components.Text",
                    ),
                }: ListArray
                [
                  StringArray
                [
                  "/test/imu",
                ],
                ],
                ComponentDescriptor {
                    archetype: Some(
                        "rerun.archetypes.McapSchema",
                    ),
                    component: "McapSchema:id",
                    component_type: Some(
                        "rerun.components.SchemaId",
                    ),
                }: ListArray
                [
                  PrimitiveArray<UInt16>
                [
                  1,
                ],
                ],
                ComponentDescriptor {
                    archetype: Some(
                        "rerun.archetypes.McapChannel",
                    ),
                    component: "McapChannel:id",
                    component_type: Some(
                        "rerun.components.ChannelId",
                    ),
                }: ListArray
                [
                  PrimitiveArray<UInt16>
                [
                  1,
                ],
                ],
            },
        ),
    },
    Chunk {
        id: ChunkId(
            0000000000661EFDf2e3b19f7c045f1a,
        ),
        entity_path: /__properties,
        heap_size_bytes: 2968,
        is_sorted: true,
        row_ids: FixedSizeBinaryArray<16>
        [
          [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
        ],
        ],
        timelines: {},
        components: ChunkComponents(
            {
                ComponentDescriptor {
                    archetype: Some(
                        "rerun.archetypes.McapStatistics",
                    ),
                    component: "McapStatistics:message_count",
                    component_type: Some(
                        "rerun.components.Count",
                    ),
                }: ListArray
                [
                  PrimitiveArray<UInt64>
                [
                  1,
                ],
                ],
                ComponentDescriptor {
                    archetype: Some(
                        "rerun.archetypes.McapStatistics",
                    ),
                    component: "McapStatistics:message_end_time",
                    component_type: Some(
                        "rerun.components.Timestamp",
                    ),
                }: ListArray
                [
                  PrimitiveArray<Int64>
                [
                  1000000000,
                ],
                ],
                ComponentDescriptor {
                    archetype: Some(
                        "rerun.archetypes.McapStatistics",
                    ),
                    component: "McapStatistics:metadata_count",
                    component_type: Some(
                        "rerun.components.Count",
                    ),
                }: ListArray
                [
                  PrimitiveArray<UInt64>
                [
                  0,
                ],
                ],
                ComponentDescriptor {
                    archetype: Some(
                        "rerun.archetypes.McapStatistics",
                    ),
                    component: "McapStatistics:channel_message_counts",
                    component_type: Some(
                        "rerun.components.ChannelMessageCounts",
                    ),
                }: ListArray
                [
                  ListArray
                [
                  StructArray
                -- validity:
                [
                  valid,
                ]
                [
                -- child 0: "channel_id" (UInt16)
                PrimitiveArray<UInt16>
                [
                  1,
                ]
                -- child 1: "message_count" (UInt64)
                PrimitiveArray<UInt64>
                [
                  1,
                ]
                ],
                ],
                ],
                ComponentDescriptor {
                    archetype: Some(
                        "rerun.archetypes.McapStatistics",
                    ),
                    component: "McapStatistics:channel_count",
                    component_type: Some(
                        "rerun.components.Count",
                    ),
                }: ListArray
                [
                  PrimitiveArray<UInt64>
                [
                  1,
                ],
                ],
                ComponentDescriptor {
                    archetype: Some(
                        "rerun.archetypes.McapStatistics",
                    ),
                    component: "McapStatistics:chunk_count",
                    component_type: Some(
                        "rerun.components.Count",
                    ),
                }: ListArray
                [
                  PrimitiveArray<UInt64>
                [
                  1,
                ],
                ],
                ComponentDescriptor {
                    archetype: Some(
                        "rerun.archetypes.McapStatistics",
                    ),
                    component: "McapStatistics:attachment_count",
                    component_type: Some(
                        "rerun.components.Count",
                    ),
                }: ListArray
                [
                  PrimitiveArray<UInt64>
                [
                  0,
                ],
                ],
                ComponentDescriptor {
                    archetype: Some(
                        "rerun.archetypes.McapStatistics",
                    ),
                    component: "McapStatistics:message_start_time",
                    component_type: Some(
                        "rerun.components.Timestamp",
                    ),
                }: ListArray
                [
                  PrimitiveArray<Int64>
                [
                  1000000000,
                ],
                ],
                ComponentDescriptor {
                    archetype: Some(
                        "rerun.archetypes.McapStatistics",
                    ),
                    component: "McapStatistics:schema_count",
                    component_type: Some(
                        "rerun.components.Count",
                    ),
                }: ListArray
                [
                  PrimitiveArray<UInt64>
                [
                  1,
                ],
                ],
            },
        ),
    },
]
